Packages/com.unity.inputsystem/InputSystem/Editor/PropertyDrawers/InputActionReferencePropertyDrawer.cs

```diff
@@ -0,0 +1,47 @@
+// Note: If not UNITY_INPUT_SYSTEM_PROJECT_WIDE_ACTIONS we do not use a custom property drawer and
+//       picker for InputActionReferences but rather rely on default (classic) object picker.
+#if UNITY_EDITOR && UNITY_INPUT_SYSTEM_PROJECT_WIDE_ACTIONS
+
+using UnityEditor;
+using UnityEditor.Search;
+
+namespace UnityEngine.InputSystem.Editor
+{
+    /// <summary>
+    /// Custom property drawer in order to use the "Advanced Picker" from UnityEditor.Search.
+    /// </summary>
+    [CustomPropertyDrawer(typeof(InputActionReference))]
+    internal sealed class InputActionReferencePropertyDrawer : PropertyDrawer
+    {
+        private readonly SearchContext m_Context = UnityEditor.Search.SearchService.CreateContext(new[]
+        {
+            InputActionReferenceSearchProviders.CreateInputActionReferenceSearchProviderForAssets(),
+            InputActionReferenceSearchProviders.CreateInputActionReferenceSearchProviderForProjectWideActions(),
+        }, string.Empty, SearchConstants.PickerSearchFlags);
+
+
+        public override void OnGUI(Rect position, SerializedProperty property, GUIContent label)
+        {
+            // Sets the property to null if the action is not found in the asset.
+            ValidatePropertyWithDanglingInputActionReferences(property);
+
+            ObjectField.DoObjectField(position, property, typeof(InputActionReference), label,
+                m_Context, SearchConstants.PickerViewFlags);
+        }
+
+        static void ValidatePropertyWithDanglingInputActionReferences(SerializedProperty property)
+        {
+            if (property?.objectReferenceValue is InputActionReference reference)
+            {
+                var action = reference?.asset?.FindAction(reference.action.id);
+                if (action is null)
+                {
+                    property.objectReferenceValue = null;
+                    property.serializedObject.ApplyModifiedProperties();
+                }
+            }
```

Comment:
Perhaps here we also need to refresh the InputActionReference reference name in case its changed in the action.

Reply:
No need.
When renaming an input action, we need to manually recreate the action reference in the asset ([code here](https://github.com/Unity-Technologies/InputSystem/blob/f23f58720ff7dd5e7678ffb3469e999a61325f5a/Packages/com.unity.inputsystem/InputSystem/Editor/ProjectWideActions/ProjectWideActionsAsset.cs#L145-L150)). Otherwise, the action reference serialized object in the .asset file has the name of an action that no longer exists.
At the same time, the change in name is reflected in the inspector UI of this property drawer.